Sources/AblyAssetTrackingPublisher/DefaultPublisher.swift

```diff
@@ -212,34 +212,8 @@ extension DefaultPublisher {
 
     // MARK: Track
     private func performTrackTrackableEvent(_ event: Event.TrackTrackableEvent) {
-        guard !state.isStoppingOrStopped else {
-            publisherStoppedCallback(handler: event.resultHandler)
-            return
-        }
-
-        guard !trackables.contains(event.trackable) else {
-            enqueue(event: .trackableReadyToTrack(.init(trackable: event.trackable, resultHandler: event.resultHandler)))
-            return
-        }
-
-        ablyPublisher.connect(
-            trackableId: event.trackable.id,
-            presenceData: presenceData,
-            useRewind: false
-        ) { [weak self] result in
-            switch result {
-            case .success:
-                self?.enqueue(event: .presenceJoinedSuccessfully(.init(trackable: event.trackable) { [weak self] result in
-                    switch result {
-                    case .success:
-                        self?.enqueue(event: .trackableReadyToTrack(.init(trackable: event.trackable, resultHandler: event.resultHandler)))
-                    default:
-                        return
-                    }
-                }))
-            case .failure(let error):
-                self?.callback(error: error, handler: event.resultHandler)
-            }
+        performAddOrTrack(event.trackable, resultHandler: event.resultHandler) {
+            self.enqueue(event: .trackableReadyToTrack(.init(trackable: event.trackable, resultHandler: event.resultHandler)))
```

Comment:
Do we want to enqueue `trackableReadyToTrack` even if `performAddOrTrack` failed?
Earlier we did it only if `presenceJoinedSuccessfully` happened:
```
self?.enqueue(event: .presenceJoinedSuccessfully(.init(trackable: event.trackable) { [weak self] result in
                    switch result {
                    case .success:
                        self?.enqueue(event: .trackableReadyToTrack(.init(trackable: event.trackable, resultHandler: event.resultHandler)))
                    default:
                        return
                    }
                }))
```

Reply:
This should have been dealt with inside `performAddOrTrack`

If you check this code block
```swift
 switch result {
            case .success:
                self?.enqueue(event: .presenceJoinedSuccessfully(.init(trackable: trackable) { [weak self] _ in
                    self?.callback(value: Void(), handler: resultHandler)
                }))
            case .failure(let error):
                self?.callback(error: error, handler: resultHandler)
            }
```

Reply:
I might be reading this wrong, it's been a long day already, but I think it also won't be invoked when there were no issues and `presenceJoinedSuccessfully`, since you're not calling `completion()` here, only the `callback`:
```swift
                self?.enqueue(event: .presenceJoinedSuccessfully(.init(trackable: trackable) { [weak self] _ in
                    self?.callback(value: Void(), handler: resultHandler)
                }))
```

So currently, if I understand this correctly, the `. trackableReadyToTrack` will be enqueued only if this guard fails (so if a trackable is already added):
```swift
        guard !trackables.contains(trackable) else {
            completion()
            return
        }
```

Reply:
So in essence - I think you should just be calling the completion here instead of a callback:
```swift
self?.enqueue(event: .presenceJoinedSuccessfully(.init(trackable: trackable) { [weak self] _ in
                    completion()
                }))
```

Reply:
Of course it's here: 5c4b4986836d7ab92e8755c7f0cc21bed699148a. I'm also brain dead right now but I'd expect there would be. a test coverage for this as this is just refactoring